src/Type/IntersectionType.php

```diff
@@ -0,0 +1,188 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type;
+
+use PHPStan\TrinaryLogic;
+
+class IntersectionType implements CompoundType, StaticResolvableType
+{
+
+	/** @var \PHPStan\Type\Type[] */
+	private $types;
+
+	/**
+	 * @param Type[] $types
+	 */
+	public function __construct(array $types)
+	{
+		$this->types = $types;
+	}
+
+	/**
+	 * @return Type[]
+	 */
+	public function getTypes(): array
+	{
+		return $this->types;
+	}
+
+	/**
+	 * @return string|null
+	 */
+	public function getClass()
+	{
+		return UnionTypeHelper::getClass($this->types);
+	}
+
+	/**
+	 * @return string[]
+	 */
+	public function getReferencedClasses(): array
+	{
+		return UnionTypeHelper::getReferencedClasses($this->types);
+	}
+
+	public function combineWith(Type $otherType): Type
+	{
+		return TypeCombinator::union($this, $otherType);
+	}
+
+	public function accepts(Type $otherType): bool
+	{
+		foreach ($this->types as $type) {
+			if (!$type->accepts($otherType)) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	public function isSupersetOf(Type $otherType): TrinaryLogic
+	{
+		$results = [];
+		foreach ($this->getTypes() as $innerType) {
+			$results[] = $innerType->isSupersetOf($otherType);
+		}
+
+		return TrinaryLogic::createYes()->and(...$results);
+	}
+
+	public function isSubsetOf(Type $otherType): TrinaryLogic
+	{
+		if ($otherType instanceof self || $otherType instanceof UnionType) {
+			return $otherType->isSupersetOf($this);
+		}
+
+		$results = [];
+		foreach ($this->getTypes() as $innerType) {
+			$results[] = $otherType->isSupersetOf($innerType);
+		}
+
+		return TrinaryLogic::maxMin(...$results);
+	}
+
+	public function describe(): string
+	{
+		/** @var IterableIterableType[] $iterableTypes */
+		$iterableTypes = [];
+
+		/** @var Type[] $otherTypes */
+		$otherTypes = [];
+
+		foreach ($this->types as $type) {
+			if ($type instanceof IterableIterableType) {
+				$iterableTypes[] = $type;
+			} else {
+				$otherTypes[] = $type;
+			}
+		}
+
+		$typeNames = [];
+
+		foreach (UnionTypeHelper::sortTypes($this->types) as $type) {
+			$typeNames[] = $type->describe();
+		}
+
+		return '(' . implode('&', $typeNames) . ')';
```

Comment:
I don't think it always has to be in parentheses. I think they should be added only if there are nested types. Also, the same logic should probably be in the UnionType - add parentheses only if the meaning is ambiguous.